File: stitch_vid.py
import cv2
import numpy as np
import time
import os

# Remove previous keyframes
os.system('rm ./keyframes/*')

# Load video file
cap = cv2.VideoCapture('input_home1.mp4')

# Extract keyframes
start_time = time.time()
keyframes = []
num_keyframes = 200
keytry_cnt = 0
while len(keyframes) < num_keyframes:
    ret, frame = cap.read()
    if not ret:
        break
    # Compute frame difference
    if len(keyframes) > 0:
        diff = cv2.absdiff(frame, keyframes[-1])
        mean_diff = np.mean(diff)
    # Compute focus measure using Laplacian variance
    fm = cv2.Laplacian(frame, cv2.CV_64F).var()
    # Add frame as keyframe if it's visually different enough from previous keyframes and has sufficient focus
    if len(keyframes) == 0 or (mean_diff > 40 and fm > 300):
        keyframes.append(frame)

extraction_time = time.time() - start_time

# Save keyframes as JPEG files
start_time = time.time()
for i, frame in enumerate(keyframes):
    cv2.imwrite("keyframes/keyframe{}.jpg".format(i), frame)
    print("Keyframe {} saved as keyframe{}.jpg".format(i, i))
save_time = time.time() - start_time

# Stitch keyframes
stitcher = cv2.createStitcher() if cv2.__version__.startswith('3') else cv2.Stitcher_create()
status, stitched = stitcher.stitch(keyframes)
stitching_time = time.time() - start_time

# Save result as JPEG file
if status == cv2.STITCHER_OK:
    cv2.imwrite("output.jpg", stitched)
    print("Stitched panorama saved as output.jpg")
else:
    print("Error stitching images:", status)
save_time = time.time() - start_time

# Print execution times
print("Keyframe extraction time:", extraction_time, "seconds")
print("Saving keyframes time:", save_time, "seconds")
print("Stitching time:", stitching_time, "seconds")
print("Saving stitched panorama time:", save_time, "seconds")


### Display result
##if status == cv2.STITCHER_OK:
##    cv2.imshow("Panorama", stitched)
##    cv2.waitKey(0)
##    cv2.destroyAllWindows()
##else:
##    print("Error stitching images:", status)

# Release resources
cap.release()

File: stitch_vid.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <unistd.h>

using namespace cv;
using namespace std;

int main()
{
    // Remove previous keyframes
    system("rm ./keyframes/*");

    // Load video file
    VideoCapture cap("input_home1.mp4");

    // Extract keyframes
    int num_keyframes = 200;
    vector<Mat> keyframes;
    int keytry_cnt = 0;
    while (keyframes.size() < num_keyframes)
    {
        Mat frame;
        cap >> frame;
        if (frame.empty())
        {
            break;
        }
        // Compute frame difference
        double mean_diff;
        if (!keyframes.empty())
        {
            Mat diff;
            absdiff(frame, keyframes.back(), diff);
            Scalar mean_val = mean(diff);
            mean_diff = mean_val[0] + mean_val[1] + mean_val[2];
            mean_diff /= 3.0;
        }
        else
        {
            mean_diff = 0.0;
        }
        // Compute focus measure using Laplacian variance
        Mat gray_frame;
        cvtColor(frame, gray_frame, COLOR_BGR2GRAY);
        Mat lap_frame;
        Laplacian(gray_frame, lap_frame, CV_64F);
        Scalar lap_var = mean(lap_frame.mul(lap_frame));
        double fm = lap_var[0];

        // Add frame as keyframe if it's visually different enough from previous keyframes and has sufficient focus
        if (keyframes.empty() || (mean_diff > 40.0 && fm > 300.0))
        {
            keyframes.push_back(frame);
            imwrite("keyframes/keyframe" + to_string(keyframes.size() - 1) + ".jpg", frame);
            cout << "Keyframe " << keyframes.size() - 1 << " saved as keyframe" << keyframes.size() - 1 << ".jpg" << endl;
        }
    }

    // Stitch keyframes
    Stitcher::Mode mode = Stitcher::PANORAMA;
    Ptr<Stitcher> stitcher = Stitcher::create(mode);
    Stitcher::Status status;
    Mat stitched;
    status = stitcher->stitch(keyframes, stitched);

    // Save result as JPEG file
    if (status == Stitcher::OK)
    {
        imwrite("output.jpg", stitched);
        cout << "Stitched panorama saved as output.jpg" << endl;
    }
    else
    {
        cout << "Error stitching images: " << status << endl;
    }

    // Print execution times
    double extraction_time, save_time, stitching_time;
    clock_t start_time = clock();
    extraction_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
    start_time = clock();
    save_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
    start_time = clock();
    stitching_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
    cout << "Keyframe extraction time: " << extraction_time << " seconds" << endl;
    cout << "Saving keyframes time: " << save_time << " seconds" << endl;
    cout << "Stitching time: " << stitching_time << " seconds" << endl;
    cout << "Saving stitched panorama time: " << save_time << " seconds" << endl;

    // Release resources
    cap.release();

    return
}
